### ISSUE TITLE
Sequence parsers do not roll back when some sub-parsers fail

### ISSUE THREAD

**Issue by postrov** [a-h/parse#14]
Say we use a SequenceOf2 parser. When first sub-parer matches, but second does not, the whole SequenceOf2 parser does not match (as expected). However, the input is not rolled back and is left at the position after first sub-parser. I created a minimal example showing this:

```go
package main

import (
	"fmt"

	"github.com/a-h/parse"
)

var firstMatchButSecondNoMatch = parse.SequenceOf2[string, string](
	parse.Rune('A'),
	parse.Rune('B'),
)

var bothMatch = parse.SequenceOf2[string, string](
	parse.Rune('A'),
	parse.Rune('C'),
)

// this would match if we swapped the order of parsers in Any
var tryBoth = parse.Any[parse.Tuple2[string, string]](
	firstMatchButSecondNoMatch,
	bothMatch,
)

func main() {
	input := parse.NewInput("AC")

	match, ok, err := tryBoth.Parse(input)

	fmt.Println(match, ok, err) // { } false <nil>, expected to match on bothMatch parser
	fmt.Println(input.Index())  // 1, expected to be still at 0 after failed match
}
```

**Comment by postrov** [comment]
I guess remembering start position and seeking to it on sub-parser failure would fix this, much like in the `All` parser.

**Comment by postrov** [comment]
Created a PR for that: https://github.com/a-h/parse/pull/15

**Comment by a-h** [comment]
Fixed in #15 - thanks again! Nice work. 😀
